packages/agent-base/src/index.ts

```diff
@@ -77,6 +78,61 @@ export abstract class Agent extends http.Agent {
 			);
 	}
 
+	// In order to support async signatures in `connect()` and Node's native
+	// connection pooling in `http.Agent`, the array of sockets for each origin
+	// has to be updated synchronously. This is so the length of the array is
+	// accurate when `addRequest()` is next called. We achieve this by creating a
+	// fake socket and adding it to `sockets[origin]` and incrementing
+	// `totalSocketCount`.
+	private incrementSockets(name: string) {
+		// If `maxSockets` and `maxTotalSockets` are both Infinity then there is no
+		// need to create a fake socket because Node.js native connection pooling
+		// will never be invoked.
+		if (this.maxSockets === Infinity && this.maxTotalSockets === Infinity) {
+			return null;
+		}
+		// All instances of `sockets` are expected TypeScript errors. The
+		// alternative is to add it as a private property of this class but that
+		// will break TypeScript subclassing.
+		if (!this.sockets[name]) {
+			// @ts-expect-error `sockets` is readonly in `@types/node`
+			this.sockets[name] = [];
+		}
+		const fakeSocket = new net.Socket({ writable: false });
+		(this.sockets[name] as net.Socket[]).push(fakeSocket);
+		// @ts-expect-error `totalSocketCount` isn't defined in `@types/node`
+		this.totalSocketCount++;
+		return fakeSocket;
+	}
+
+	private decrementSockets(name: string, socket: null | net.Socket) {
+		if (!this.sockets[name] || socket === null) {
+			return;
+		}
+		const sockets = this.sockets[name] as net.Socket[];
+		const index = sockets.indexOf(socket);
+		if (index !== -1) {
+			sockets.splice(index, 1);
+			// @ts-expect-error  `totalSocketCount` isn't defined in `@types/node`
+			this.totalSocketCount--;
+			if (sockets.length === 0) {
+				// @ts-expect-error `sockets` is readonly in `@types/node`
+				delete this.sockets[name];
+			}
+		}
+	}
+
+	// In order to properly update the socket pool, we need to call `getName()` on
+	// the core `https.Agent` if it is a secureEndpoint.
+	private getName({ secureEndpoint, ...options }: AgentConnectOpts) {
```

Comment:
Maybe this should be `public`? Since `agent-base` only subclasses `http.Agent` I think there are other times when it might be useful to users to call `agent.getName()` and get the [Node https.Agent behavior](https://github.com/nodejs/node/blob/v20.11.1/lib/https.js#L233-L318)

Reply:
Ya that would be good I think.

Also, I've had luck in the past sending PRs to DefinitelyTyped. We could clean up some of these `@ts-expect-error`s if we added the missing bits.

Reply:
That's a good idea. I think at least `getName()`, `addRequest()`, and `createSocket()` would benefit from being added to DefinitelyTyped.

I think `sockets` would probably stay as `readonly` and if `totalSocketCount` was added that should be `readonly` too. Even though this PR writes to them, it's not exactly encouraged behavior.

Reply:
Updated `getName` to be public and also not require `secureEndpoint` to be passed in. If it's not a boolean then it will fallback to calling `this.isSecureEndpoint(opts)`.